File: typescript/tracing/mcp-openinference/mcp_openinference_example.ts
/** MCP — Client calling a tool on a stdio server, traced via OpenInference.
 *
 * Both client and server are instrumented with @arizeai/openinference-instrumentation-mcp.
 * The MCP instrumentation propagates W3C trace context through JSON-RPC _meta,
 * linking client and server spans into a single distributed trace.
 */

import "dotenv/config";
import { trace } from "@opentelemetry/api";
import { MCPInstrumentation } from "@arizeai/openinference-instrumentation-mcp";
import { Respan, withWorkflow, withTool } from "@respan/respan";
import { OpenInferenceInstrumentor } from "@respan/instrumentation-openinference";

const RESPAN_API_KEY = process.env.RESPAN_API_KEY;
const RESPAN_BASE_URL = (process.env.RESPAN_BASE_URL ?? "https://api.respan.ai/api").replace(/\/+$/, "");

const respan = new Respan({
  apiKey: RESPAN_API_KEY,
  baseURL: RESPAN_BASE_URL,
  instrumentations: [new OpenInferenceInstrumentor(MCPInstrumentation)],
});
await respan.initialize();

const MCPClientModule = await import("@modelcontextprotocol/sdk/client/index.js");
const MCPClientStdioModule = await import("@modelcontextprotocol/sdk/client/stdio.js");

const mcpInst = new MCPInstrumentation();
mcpInst.setTracerProvider(trace.getTracerProvider());
mcpInst.manuallyInstrument({ clientStdioModule: MCPClientStdioModule });

const transport = new MCPClientStdioModule.StdioClientTransport({
  command: "npx",
  args: ["tsx", new URL("./server.ts", import.meta.url).pathname],
  env: { ...process.env } as Record<string, string>,
});

const client = new MCPClientModule.Client({ name: "test-client", version: "1.0.0" });

await withWorkflow({ name: "mcp-hello-workflow" }, async () => {
  await client.connect(transport);

  const tools = await withTool({ name: "mcp.listTools" }, async () => {
    return client.listTools();
  });
  console.log("Available tools:", tools.tools.map((t: any) => t.name));

  const result = await withTool({ name: "mcp.callTool:hello" }, async () => {
    return client.callTool({ name: "hello", arguments: { name: "Respan" } });
  });
  console.log("Tool result:", (result.content as any)[0].text);

  await client.close();
});

await respan.flush();

File: python/tracing/mcp-openinference/mcp_openinference_example.py
#!/usr/bin/env python3
"""
MCP Client + OpenInference example for Respan tracing.

Uses openinference-instrumentation-mcp to propagate OTel context across
the MCP stdio transport boundary. The client creates a workflow span,
connects to an MCP server subprocess, and calls a tool. Because of the
instrumentation, the server's spans join the client's trace.
"""
import asyncio
import os
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT_ENV_PATH = Path(__file__).resolve().parents[3] / ".env"
load_dotenv(ROOT_ENV_PATH, override=True)

RESPAN_API_KEY = os.getenv("RESPAN_GATEWAY_API_KEY") or os.getenv("RESPAN_API_KEY")
RESPAN_BASE_URL = (
    os.getenv("RESPAN_GATEWAY_BASE_URL")
    or os.getenv("RESPAN_BASE_URL")
    or "https://api.respan.ai/api"
).rstrip("/")

from openinference.instrumentation.mcp import MCPInstrumentor
from respan_instrumentation_openinference import OpenInferenceInstrumentor
from respan_tracing import RespanTelemetry
from respan_tracing.decorators import workflow

telemetry = RespanTelemetry(
    app_name="mcp-openinference-example",
    api_key=RESPAN_API_KEY,
    is_batching_enabled=False,
    instruments=set(),
)

mcp_oi = OpenInferenceInstrumentor(MCPInstrumentor)
mcp_oi.activate()

from mcp.client.stdio import stdio_client
from mcp.client.session import ClientSession
from mcp import StdioServerParameters


@workflow(name="mcp_openinference_workflow")
async def mcp_workflow() -> str:
    server_script = str(Path(__file__).parent / "mcp_server.py")
    server_params = StdioServerParameters(
        command=sys.executable,
        args=[server_script],
    )

    async with stdio_client(server_params) as (read_stream, write_stream):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()

            tools = await session.list_tools()
            print(f"Available tools: {[t.name for t in tools.tools]}")

            result = await session.call_tool("get_haiku", {"topic": "recursion in programming"})
            text = result.content[0].text if result.content else ""
            return text


async def main() -> None:
    print("=" * 60)
    print("MCP OpenInference Example")
    print("=" * 60)

    if not RESPAN_API_KEY:
        print("Skipping: RESPAN_API_KEY not set.")
        return

    try:
        result = await mcp_workflow()
        print(result)
    finally:
        telemetry.flush()
        mcp_oi.deactivate()


if __name__ == "__main__":
    asyncio.run(main())
